kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt

```diff
@@ -0,0 +1,759 @@
+    // This is a "fake" class and not visible to Elements
+    return true
+  }
+  return elementUtils.getTypeElement(fqcn).kind == ElementKind.INTERFACE
+}
+
+private fun TypeName.objectType(): TypeName {
+  return when (this) {
+    BOOLEAN -> Boolean::class.javaObjectType.asTypeName()
+    BYTE -> Byte::class.javaObjectType.asTypeName()
+    SHORT -> Short::class.javaObjectType.asTypeName()
+    INT -> Integer::class.javaObjectType.asTypeName()
+    LONG -> Long::class.javaObjectType.asTypeName()
+    CHAR -> Character::class.javaObjectType.asTypeName()
+    FLOAT -> Float::class.javaObjectType.asTypeName()
+    DOUBLE -> Double::class.javaObjectType.asTypeName()
+    else -> this
+  }
+}
+
+private fun TypeName.makeType(
+    elementUtils: Elements,
+    typesArray: ParameterSpec,
+    genericTypeNames: List<TypeVariableName>): CodeBlock {
+  if (nullable) {
+    return asNonNullable().makeType(elementUtils, typesArray, genericTypeNames)
+  }
+  return when (this) {
+    is ClassName -> CodeBlock.of("%T::class.java", this)
+    is ParameterizedTypeName -> {
+      // If it's an Array type, we shortcut this to return Types.arrayOf()
+      if (rawType == ARRAY) {
+        return CodeBlock.of("%T.arrayOf(%L)",
+            Types::class.asTypeName(),
+            typeArguments[0].objectType().makeType(elementUtils, typesArray, genericTypeNames))
+      }
+      // If it's a Class type, we have to specify the generics.
+      val rawTypeParameters = if (rawType.isClass(elementUtils)) {
+        CodeBlock.of(
+            typeArguments.joinTo(
+                buffer = StringBuilder(),
+                separator = ", ",
+                prefix = "<",
+                postfix = ">") { "%T" }
+                .toString(),
+            *(typeArguments.map { objectType() }.toTypedArray())
+        )
+      } else {
+        CodeBlock.of("")
+      }
+      CodeBlock.of(
+          "%T.newParameterizedType(%T%L::class.java, ${typeArguments
+              .joinToString(", ") { "%L" }})",
+          Types::class.asTypeName(),
+          rawType.objectType(),
+          rawTypeParameters,
+          *(typeArguments.map {
+            it.objectType().makeType(elementUtils, typesArray, genericTypeNames)
+          }.toTypedArray()))
+    }
+    is WildcardTypeName -> {
+      val target: TypeName
+      val method: String
+      when {
+        lowerBounds.size == 1 -> {
+          target = lowerBounds[0]
+          method = "supertypeOf"
+        }
+        upperBounds.size == 1 -> {
+          target = upperBounds[0]
+          method = "subtypeOf"
+        }
+        else -> throw IllegalArgumentException(
+            "Unrepresentable wildcard type. Cannot have more than one bound: " + this)
+      }
+      CodeBlock.of("%T.%L(%T::class.java)", Types::class.asTypeName(), method, target)
+    }
+    is TypeVariableName -> {
+      CodeBlock.of("%N[%L]", typesArray, genericTypeNames.indexOfFirst { it == this })
+    }
+  // Shouldn't happen
```

Comment:
Remove. The exception already says this.

Reply:
0702fc7